src/mcp/server/fastmcp/server.py

```diff
@@ -782,16 +782,21 @@ def streamable_http_app(self) -> Starlette:
             )
 
         # Create the ASGI handler
-        async def handle_streamable_http(
-            scope: Scope, receive: Receive, send: Send
-        ) -> None:
-            await self.session_manager.handle_request(scope, receive, send)
+        async def handle_streamable_http(request: Request) -> Response:
+            await self.session_manager.handle_request(
+                request.scope, request.receive, request._send
+            )
+            return Response()
 
         # Create routes
-        routes: list[Route | Mount] = []
+        routes: list[Route] = []
         middleware: list[Middleware] = []
         required_scopes = []
 
+        # Always register both /mcp and /mcp/ for full compatibility
+        _main_path = self.settings.streamable_http_path.removesuffix("/")
+        _alt_path = _main_path + "/"
```

Comment:
I think the problem was the redirect, but if you have a 404 then it's not an issue anymore, and we shouldn't be needing this.